lib/client/db/mcp/server.go

```diff
@@ -0,0 +1,135 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package mcp
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+
+	"github.com/ghodss/yaml"
+	"github.com/gravitational/trace"
+	"github.com/mark3labs/mcp-go/mcp"
+	mcpserver "github.com/mark3labs/mcp-go/server"
+
+	"github.com/gravitational/teleport"
+)
+
+// listDatabasesTool is the MCP tool that list all databases being served
+// (from all protocols).
+var listDatabasesTool = mcp.NewTool(listDatabasesToolName,
+	mcp.WithDescription("List database resources available to be used with Teleport tools."),
+)
+
+// RootServer database access root MCP server.
+type RootServer struct {
+	*mcpserver.MCPServer
+	logger             *slog.Logger
+	availableDatabases map[string]*Database
+}
+
+// NewRootServer initializes a new root MCP server.
+func NewRootServer(logger *slog.Logger) *RootServer {
+	server := &RootServer{
+		MCPServer:          mcpserver.NewMCPServer(serverName, teleport.Version),
+		logger:             logger,
+		availableDatabases: make(map[string]*Database),
+	}
+	server.AddTool(listDatabasesTool, server.ListDatabases)
+
+	return server
+}
+
+// ListDatabases tool function used to list all available/served databases.
+func (s *RootServer) ListDatabases(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+	var res []mcp.Content
+	for _, db := range s.availableDatabases {
+		contents, err := encodeDatabaseResource(db)
+		if err != nil {
+			s.logger.ErrorContext(ctx, "error while list databases", "error", err)
+			return mcp.NewToolResultError(FormatErrorMessage(nil, err).Error()), nil
+		}
+		res = append(res, mcp.EmbeddedResource{Type: "resource", Resource: contents})
+	}
+
+	return &mcp.CallToolResult{
+		Content: res,
+	}, nil
+}
+
+// GetDatabaseResource resource handler for databases.
+func (s *RootServer) GetDatabaseResource(ctx context.Context, request mcp.ReadResourceRequest) ([]mcp.ResourceContents, error) {
+	db, ok := s.availableDatabases[request.Params.URI]
+	if !ok {
+		return nil, trace.NotFound("Database is %q not available as MCP resource", request.Params.URI)
+	}
+
+	encodedDb, err := encodeDatabaseResource(db)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return []mcp.ResourceContents{encodedDb}, nil
+}
+
+// RegisterDatabase register a database on the root server. This make it
+// available as a MCP resource.
+//
+// TODO(gabrielcorado): support dynamically registering/deregistering databases
+// after the server starts.
+func (s *RootServer) RegisterDatabase(db *Database) {
+	uri := db.ResourceURI()
+	s.availableDatabases[uri] = db
```

Comment:
Should `s.availableDatabases` be protected by a mutex? Right now it just runs locally but in case we reuse it on the server side.

Or at least add a comment explicitly stating why it's safe not to.